dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handling/DataAggregator.java

```diff
@@ -0,0 +1,202 @@
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.hisp.dhis.analytics.DataQueryParams.DX_INDEX;
+import static org.hisp.dhis.analytics.DataQueryParams.VALUE_ID;
+import static org.hisp.dhis.analytics.DataQueryParams.newBuilder;
+import static org.hisp.dhis.analytics.ProcessingHint.SINGLE_INDICATOR_REPORTING_RATE_FILTER_ITEM;
+import static org.hisp.dhis.analytics.SortOrder.ASC;
+import static org.hisp.dhis.common.DimensionalObject.DATA_X_DIM_ID;
+
+import javax.annotation.PostConstruct;
+
+import org.hisp.dhis.analytics.DataQueryParams;
+import org.hisp.dhis.common.DimensionalObject;
+import org.hisp.dhis.common.Grid;
+import org.hisp.dhis.system.grid.ListGrid;
+import org.springframework.stereotype.Component;
+
+/**
+ * Class responsible for aggregating all necessary data, provided by the
+ * handlers, into the Grid object
+ */
+@Component
+public class DataAggregator
+{
+
+    private final HeaderHandler headerHandler;
+
+    private final MetadataHandler metaDataHandler;
+
+    private final DataHandler dataHandler;
+
+    public DataAggregator( HeaderHandler headerHandler, MetadataHandler metadataHandler, DataHandler dataHandler )
+    {
+        checkNotNull( headerHandler );
+        checkNotNull( metadataHandler );
+        checkNotNull( dataHandler );
+
+        this.headerHandler = headerHandler;
+        this.metaDataHandler = metadataHandler;
+        this.dataHandler = dataHandler;
+    }
+
+    /**
+     * Returns a grid with aggregated data.
+     *
+     * @param params the {@link DataQueryParams}.
+     * @return a grid with aggregated data.
+     */
+    public Grid getAggregatedDataValueGridInternal( DataQueryParams params )
```

Comment:
I would consider removing the "Internal" wording, since I think it made sense when part of a larger class but in this new `Aggregator` can be confusing.

Reply:
Yes, I can do it

Reply:
Done